server/src/main/java/org/apache/iotdb/db/mpp/plan/execution/QueryExecution.java

```diff
@@ -161,11 +161,13 @@ public void start() {
       stateMachine.transitionToRunning();
       return;
     }
+    long startTime = System.currentTimeMillis();
     doLogicalPlan();
     doDistributedPlan();
     if (context.getQueryType() == QueryType.READ) {
       initResultHandle();
     }
+    context.setTimeOut(context.getTimeOut() - (System.currentTimeMillis() - startTime));
```

Comment:
We should record the original timeout and startTime in context, and then just before dispatching we calculating the exact remaining time and set it to FragmentInstance. If it's already less than or equals to 0, just throw timeout exception instead of keep dispatching.

Reply:
fixed